R/smoothDegreeDays.R

```diff
@@ -0,0 +1,131 @@
+#'   }
+#'
+#' @param nSmoothIter An integer specifying the number of iterations for lowpass smoothing.
+#' Defaults to \code{50}.
+#'
+#' @param transitionYears An integer specifying the number of years for the transition period
+#' from historical observations to projections. Defaults to \code{10}.
+#'
+#' @param nHistYears An integer specifying the number of years used for the linear regression
+#' to blend the continued historical trend into the scenario projections over a specified period
+#' in \code{transitionYears}.
+#'
+#' @return A data frame with smoothed degree day values and a seamless transition period
+#' between historical and projected data.
+#'
+#' @details
+#' The function applies lowpass filtering to smooth model projections and uses a linear
+#' interpolation to create a consistent transition period. The transition ensures that
+#' all projections align with the historical data at the transition's start, diverging smoothly
+#' into their respective future trajectories.
+#'
+#' @importFrom dplyr ungroup group_by across all_of mutate reframe filter
+#' left_join select case_when group_modify
+#' @importFrom magclass lowpass
+#' @importFrom stats lm predict
+#'
+#' @author Hagen Tockhorn
+
+
+smoothDegreeDays <- function(data, nSmoothIter = 50, transitionYears = 10, nHistYears = 10) {
+
+  # PARAMETERS -----------------------------------------------------------------
+
+  # upper temporal threshold for historical data points
+  endOfHistory <- 2020
+
+
+
+  # PROCESS DATA ---------------------------------------------------------------
+
+  dataSmooth <- data %>%
+
+    # smooth model data before averaging to avoid impact of outliers
+    group_by(across(-all_of(c("period", "value")))) %>%
+    mutate(value = ifelse(.data[["period"]] <= endOfHistory |
+                            .data[["rcp"]] == "picontrol",
+                          .data[["value"]],
+                          lowpass(.data[["value"]], i = nSmoothIter))) %>%
+    ungroup() %>%
+
+    # take mean over models
+    group_by(across(-all_of(c("model", "value")))) %>%
+    reframe(value = mean(.data[["value"]])) %>%
+    ungroup()
+
+
+  # fit linear regression to last n (= nHistYears) values per model and create
+  # predictions for m (= transitionYears) periods
+  transitionPreds <- data %>%
+    filter(.data[["period"]] >= (endOfHistory - nHistYears),
+           .data[["period"]] <= endOfHistory,
+           .data[["rcp"]] == "historical",
+           !is.na(.data[["value"]])) %>%
+
+    group_by(across(all_of(c("region", "variable", "tlim", "model")))) %>%
+    group_modify(~{
+      # create prediction data frame
+      predPeriods <- data.frame(
+        period = seq(endOfHistory, endOfHistory + transitionYears, 1)
+      )
+
+      # linear fit
+      fit <- lm(value ~ period, data = .x)
+
+      # predict future values
+      predPeriods$prediction <- predict(fit, newdata = predPeriods)
+      return(predPeriods)
+    }) %>%
```

Comment:
I Think I'd find it easier to read if you outsourced this to a subfunction. But it's okay. Matter of taste...